aws/src/main/java/org/apache/iceberg/aws/AwsClientConfigFactory.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws;
+
+import java.io.Serializable;
+import org.apache.iceberg.catalog.CatalogConfigurable;
+import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
+import software.amazon.awssdk.awscore.client.builder.AwsSyncClientBuilder;
+
+/**
+ * Interface for any customization of AWS client.
+ * <p>
+ * User can provide an implementation of this factory to perform any modification to AWS clients,
+ * such as credentials provider, region, endpoint, http client, etc.
+ * See {@link AssumeRoleClientConfigFactory} as an example.
+ */
+public interface AwsClientConfigFactory extends CatalogConfigurable, Serializable {
```

Comment:
Another thought (and I'm starting to think this may just make more sense) is to merge the idea of the `AwsClientConfigFactory` and `AwsClientUtil` and just make a single dynamically loadable class that would be a factory for creating the clients.  All the `AwsClientUtil` does at this point is instantiate the classes and all the `AwsClientConfigFactory` does is configure it, so it would be pretty sensible to just make one that does both as a `AwsClientFactory`

Then you just have 

`AwsClientFactory::<service>Client()` with a `DefaultAwsClientFactory` that is loaded if a custom implementation is not supplied.

Reply:
Yes definitely agree that factory is not a good name, and I put it in `AwsProperties::configure` just to satisfy the interface of `client.applyMutation(...)`, but I also feel it is a bit awkward. `DefaultAwsClientFactory` sounds like a good idea, let me do that.

Reply:
However, I don't want people implementing the class to keep updating when we add new AWS clients, let me think about that for a moment

Reply:
@danielcweeks So after testing a few different approaches, I think keeping a `AwsClientConfigurer` and a single `AwsClientFactory` is the most flexible approach. Users can configure all clients using `AwsClientConfigurer`, and we can introduce new clients in `AwsClientFactory` without users changing their configurer implementation. Please let me know what you think, thank you!

Reply:
@jackye1995 I like this better, but I feel like I'm not following what value the `AwsClientConfigurer` adds at this point.  It seems that `AwsClientConfigurer` interface is just a duplication of the `setDefaultsAndConfigure` in `AwsClientFactory` (they have the same signature).  Why wouldn't we just drop the `AwsClientConfigurer` and dynamically load the factory? Then if someone wants to extend the configuration, they just extend `AwsClientFactory` and override the `configure` method?  It seems like we would still be able to add new clients and the configure method would still apply to any of the standard client builder options.

The added value this provides is that since the Factory is responsible for instantiation, we can now override that behavior (e.g. cache/reuse/proxy or force new instances for instances of the client).  That may not be necessary for glue catalog as I assume there would only be one instance, but for S3FileIO it may be helpful.

Reply:
Yes I thought about just having a factory, but here are my reasons for having the additional `AwsClientConfigurer` delegation, please let me know if you think they are important or not:
1. the interface of the factory would continue to evolve as more clients are added. I think for a public interface used for dynamic loading, it is preferred to keep it stable and unchanged for most of the time, so there won't be incompatibility errors when the package versions do not match.
2. I see there are 3 layers of configuration, (1) defaults to initialize a bare-minimum client (http client impl, etc.) (2) user-specific defaults for all their clients (credentials, etc.), (3) user-specific client configurations (s3 client retry, etc.), and the intention of `setDefaultsAndConfigure` is to configure (1) and then leave (2) and (3) for customers to configure using the configurer.
3. If a user can directly load a factory, there is nothing to stop the user from overriding the specific get client methods and skip config layer 1. This also defeats the purpose of using a general `configure` method.

Reply:
@danielcweeks any thoughts on this?

Reply:
Sorry, this just dropped off my radar.

I don't actually disagree with in concept with any of the items above, but I would propose handling it differently.

Make AwsClientFactory an interface, which would just expose the client calls (e.g. `factory.s3()`, `factory.glue()`).  Create a `DefaultAwsClientFactory` that has the common configuration `setDefaultsAndConfigure`.  Create an `StsAwsClientFactory` that overrides for `setDefaultsAndConfigure` for the common functionality.

Now I can see your point in #3 that is someone were to override the client creation, they would have to directly call the `setDefaultsAndConfigure`, but I don't feel like preventing users from overriding the creation of the client is a good tradeoff for ensuring that they're consistently configured.

We could separate the instantiation and access of the client in the factory so that you can enforce the configure path, but I feel like that's just overcomplicating the situation.  Doing the above just gets us back to a more traditional, interface and implementation design that I feel will make more sense for anyone looking to provide a custom implementation.

Reply:
Thanks for the feedback, updated, please let me know if this looks good to you.